lib/explorer/polars_backend/lazy_frame.ex

```diff
@@ -29,7 +43,64 @@ defmodule Explorer.PolarsBackend.LazyFrame do
   def lazy(ldf), do: ldf
 
   @impl true
-  def collect(ldf), do: Shared.apply_dataframe(ldf, ldf, :lf_collect, [])
+  def collect(ldf) do
+    result = apply_operations(ldf)
+
+    # TODO: handle errors
+    {:ok, polars_df} = Native.lf_collect(result)
+
+    # NOTE: should we keep the groups like we do today?
+    %{ldf | data: polars_df}
+  end
+
+  defp apply_operations(ldf) do
+    ldf
+    |> LazyStack.stack_by_groups()
```

Comment:
This looks great to me! My initial suggestion is to first merge a version where we don't call `stack_by_groups`. Basically all operations become stack operations. So this:

```elixir
  def head(ldf, rows), do: Shared.apply_dataframe(ldf, ldf, :lf_head, [rows])
```

Becomes this:

```
push_operation(ldf, {:lf_head, [rows]})
```

And now, the collect is a `Enum.reduce/3` loop that converts the stack into `Shared.apply_dataframe` calls. This will solve the first part of the problem, which is to represent the operation chain in a consistent way. The next step is chunking, but then I think we can do it one by one, operation per operation.

WDYT?

Reply:
Makes sense! So we just ignore the groups when stacking, and we use the groups when collecting only, right?